File: Client/Classes/Features/Forest/Forest.cs
using Godot;
using System;
using System.Collections.Generic;
using System.Linq;

public partial class Forest : Resource, IFeature
{
    private WorldInfo worldInfo;
    private FeatureArgs args;
    private FastNoiseLite noise;
    private float threshold;
    private List<TerrainInfo> tiles = [];
    public void Set(WorldInfo worldInfo, FeatureArgs args)
    {
        this.worldInfo = worldInfo;
        this.args = args;
        noise = args.ForestNoise;
        threshold = args.Threshold;
    }

    public void Generate()
    {
        foreach (var tile in worldInfo.TerrainInfo.Values)
        {
            var v = noise.GetNoise2D(tile.PositionI.X, tile.PositionI.Y);
            if (v > threshold)
            {
                tile.TileTopType = TileUtil.TileTopType.Tree;
            }
        }
        foreach (var tile in worldInfo.TerrainInfo.Values)
        {
            if (tile.TileTopType == TileUtil.TileTopType.Tree)
            {
                var trees = tile.Neighbours.Count(n => n != null && n.TileTopType == TileUtil.TileTopType.Tree);
                if (trees >= 3)
                {
                    tile.TopTileState = TileUtil.TileTopState.Medium;
                }
                if (trees >= 4)
                {
                    tile.TopTileState = TileUtil.TileTopState.Large;
                }
            }
        }
    }

    public void Destroy()
    {
    }

    public void Update()
    {

    }
}

File: server/internal/server/objects/features/Forest.go
package features

import (
	"server/internal/server/objects"
	"server/internal/server/objects/tiles"

	"github.com/furui/fastnoiselite-go"
)

type Forest struct {
	WorldInfo *tiles.WorldInfo
	args      *tiles.FeatureArgs
	noise     fastnoiselite.FastNoiseLite
	threshold float32
	tiles     []*tiles.TerrainInfo
}

func (f *Forest) Set(world *tiles.WorldInfo, args *tiles.FeatureArgs) {
	f.WorldInfo = world
	f.args = args
	f.threshold = args.Threshold
	f.noise = args.ForestNoise
	f.tiles = make([]*tiles.TerrainInfo, 0)
}

func (f *Forest) Generate(seed uint64) {
	f.noise.Seed = int32(seed)
	f.tiles = f.tiles[:0]

	for coord, tile := range f.WorldInfo.TerrainInfo {
		v := f.noise.GetNoise2D(
			fastnoiselite.FNLfloat(coord.X),
			fastnoiselite.FNLfloat(coord.Y),
		)

		if v > float64(f.threshold) {
			tile.TopTileType = objects.Tree
			tile.TileTopState = objects.Small

			f.WorldInfo.TerrainInfo[coord] = tile
			f.tiles = append(f.tiles, tile)
		}
	}

	for _, tile := range f.tiles {
		trees := 0

		for _, neighbor := range tile.Neighbors {
			if neighbor.TopTileType == objects.Tree {
				trees++
			}
		}

		if trees >= 4 {
			tile.TileTopState = objects.Large
		} else if trees >= 3 {
			tile.TileTopState = objects.Medium
		}
	}
}
func (f *Forest) Destroy() {
}

func (f *Forest) Update() {
}
